src/dataflow-types/src/client/controller/storage.rs

```diff
@@ -264,6 +303,80 @@ impl<'a, C: Client<T>, T: Timestamp + Lattice> StorageControllerMut<'a, C, T> {
             .ok_or(StorageError::IdentifierMissing(id))
     }
 
+    /// Persist timestamp bindings updates received from ingestion workers
+    pub(super) async fn persist_timestamp_bindings(
+        &mut self,
+        feedback: &TimestampBindingFeedback<T>,
+    ) {
+        for (id, bindings) in &feedback.bindings {
+            let mut ts_binding_stash = self
+                .storage
+                .stash
+                .collection::<PartitionId, ()>(&id.to_string())
+                .unwrap();
+
+            let collection_state = self.collection_mut(*id).expect("misisng_source_id");
+
+            let upper = ts_binding_stash.upper().unwrap();
+            let since = ts_binding_stash.since().unwrap();
+            let mut seal_frontier = upper.clone();
+
+            // Here we differentialize the bindings we got from workers
```

Comment:
What would be really helpful for me, at least, is to describe this differentialization in prose! I've totally paged out the details of that conversation that we had long ago.